Improve strings rewriter for contains

ajreynol opened the pull request:
This commit improves the rewriter for string contains :
- It encapsulates some of the functions related to rewriting contains terms
- Makes stripping constant endpoints more aggressive, including related to int.to.str
- Documents some of the key functions involved

This fixes #1189

ajreynol commented:
Starexec jobs (see CVC4-100617-stringsRewContainsImprove) : 
https://www.starexec.org/starexec/secure/details/job.jsp?id=24562
https://www.starexec.org/starexec/secure/details/job.jsp?id=24564

ajreynol commented:
This is also the first step towards addressing #1180

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+        }
+      }
+      if( removeComponent ){
+        if( n1.size()==1 ){
```
Why have the special case for `n1.size() == 1`? It seems like the else branch should work in both cases.

ajreynol replied:
The distinct is that the n1.size()==1 case can be more aggressive, since in this case *all* of the string must be in n1[0], and not just a prefix.  Here is an example where we can strip the first component in the n1.size()==1 case but not in the n1.size()>1 case:

str.contains( str.to.int(x), "123a45")  ----> false

but

str.contains( str.++( str.to.int(x), y ), "123a45") -----> str.contains( str.++( str.to.int(x), y ), "123a45") 

(e.g. no change)

The reason is that y may have "a45" as a prefix.

4tXJ7f replied:
Yes, but here we are already in the `removeComponent` case and I think that the else branch does the same thing for `n1.size()==1`:
- `nb/ne.push_back( n1[0] );` is equivalent to `nb/ne.push_back( n1[index0] );` for `n1.size()==1`
- `n1.clear()` is equivalent to `n1.erase( n1.begin(), n1.begin() + 1 );` and `n1.pop_back();` for `n1.size()==1`

ajreynol replied:
ohh I see your point (I think I got the code wrong).  The reason I distinguished here is because n1.clear() may be faster than n1.erase( n1.begin(), n1.begin()+1 ), although perhaps this is paranoia.

4tXJ7f replied:
I doubt that there is any difference between the two, especially for vectors of size 1. By the way: `n1.erase( n1.begin(), n1.begin() + 1 );` can be written as just `n1.erase(n1.begin());`.

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+            }else{
+              if(n1.size()==1) {
+                removeComponent = true;
+                return true;
```
Why return here? We should signal that containment is not possible. It should be ok to use `break` here, so `n1` would be empty and since we know that `n2` is not, we can be sure that it will be rewritten to `false`.

ajreynol replied:
This has been refactored now so that its more clear, I think you're right that break was the more sensible thing to do here

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+        if( n2[index1].isConst() ){
+          CVC4::String t = n2[index1].getConst<String>();
+          const std::vector<unsigned>& tvec = t.getVec();
+          removeComponent = n1.size()>1;
```
`removeComponent = true` should be fine here.

ajreynol replied:
This needs to be n1.size()>1. 
When n1.size()==1, then by default we cannot remove the component, but removeComponent is updated true below if there is at least one non-digit.

4tXJ7f replied:
You are right, my bad. Although this makes me think: What if `tvec` is empty and `n1.size() > 1`? Then we remove the `str.to.int` even though the next component in `n2` could be a number. This might not happen in practice but still might be better to always set `removeComponent = false` and then set it to true when we see our first non-digit (this is obviously not needed/easier if we do the more aggressive rewriting that I suggested below).

ajreynol replied:
Yes, empty tvec would be a problem, but it will never happen (empty string is removed from concat's before we get to here).  I actually added an assertion !tvec.empty() in my last commit.

ajreynol replied:
Anyways, I've refactored this part quite a bit, take a look at the new version and let me know what you think.

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+          // if n1.size()==1, then if n2[index1] contains a non-digit, we can drop the entire component
+          //    e.g. str.contains( str.to.int(x), "123a45") --> false
+          // if n1.size()>1, then if n2[index1] does not contain a digit, we can drop the entire component
+          //    e.g. str.contains( str.++( str.to.int(x), y ), "abc") --> str.contains( y, "abc" )
```
We should be able to do more aggressive rewriting here (`n1.size() > 1` case):
- forward direction: if the first letter in `tvec` is not a digit -> drop the entire component
- backward direction: if the last letter in `tvec` is not a digit -> drop the entire component

ajreynol replied:
yes, I think you're right.  I'll update this.

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1394,91 +1394,66 @@ RewriteResponse TheoryStringsRewriter::preRewrite(TNode node) {
     if( node[1].isConst() ){
       CVC4::String t = node[1].getConst<String>();
-      for(unsigned i=0; i<node[0].getNumChildren(); i++){
+      for(unsigned i=1; i<(node[0].getNumChildren()-1); i++){
```
Could we add an assert that says that `node[0][0]` and `node[0][node[0].getNumChildren() - 1]` are both not const? My understanding is that this is the case because of `stripConstantEndpoints` but it would be good to have the asserts here and maybe a short comment because it is not very obvious.

ajreynol replied:
Actually I think there are cases where this assertion does not hold.

Take str.contains( str.++( "a", x, "c", y, "b" ), "ab" ).

In this case, we cannot strip the first component, nor the last component (since x may start with "b", or y may end with "a").

ajreynol replied:
Regardless, I'll add a comment to explain why we only consider the inner children.

4tXJ7f commented on src/theory/strings/theory_strings_rewriter.cpp:
```diff
@@ -1394,91 +1394,66 @@ RewriteResponse TheoryStringsRewriter::preRewrite(TNode node) {
-              }
+          //if no overlap, we can split into disjunction
+          if( t.find(s)==std::string::npos ){
+            if( s.overlap( t )==0 && t.overlap( s )==0 ){
```
Merge the two if conditions?

ajreynol replied:
Done

ajreynol commented on src/util/regexp.h:
```diff
@@ -93,17 +93,43 @@ class CVC4_PUBLIC String {
+  * then the maximal prefix of this string that is a suffix of y is of length m.
+  *
+  * For example, if x is "abcdef", then:
+  * x.overlap("aaabc") = 3
```
Fixed

ajreynol commented:
I've addressed the review and made some improvements which includes the point by Andres that:
str.contains( str.++( str.to.int(x), y ), "a123" ) ---> str.contains( y, "a123" )

ajreynol commented:
Oh btw, I will apply clang format after approved (prefer only doing this once).

ajreynol commented:
Alright I’ll refactor this to make it simpler.

On Friday, October 27, 2017, Andres Noetzli <notifications@github.com>
wrote:

> *@4tXJ7f* commented on this pull request.
> ------------------------------
>
> In src/theory/strings/theory_strings_rewriter.cpp
> <https://github.com/CVC4/CVC4/pull/1207#discussion_r147498982>:
>
> > +            if(String::isDigit(tvec[i])) {
> +              if(n1.size()>1) {
> +                removeComponent = false;
> +                break;
> +              }
> +            }else{
> +              if(n1.size()==1) {
> +                removeComponent = true;
> +                return true;
> +              }
> +            }
> +          }
> +        }
> +      }
> +      if( removeComponent ){
> +        if( n1.size()==1 ){
>
> I doubt that there is any difference between the two, especially for
> vectors of size 1. By the way: n1.erase( n1.begin(), n1.begin() + 1 );
> can be written as just n1.erase(n1.begin());.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CVC4/CVC4/pull/1207#discussion_r147498982>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC1fG4IUCdqSUC0E5FQHDsg9cw_P88eFks5swjCygaJpZM4PxKUF>
> .
>
